File: template_manager/shorthand.py
from template_manager.meta_block import MetaBlock
from template_manager.reference import Reference


class Shorthand:
    def __init__(self, type, description, references=None, meta_block=None):
        self.type = type
        self.description = description

        self.references = [Reference(reference) for reference in references] if references else []
        self.meta_block = MetaBlock(**meta_block) if meta_block else MetaBlock()

File: wuff/template/Shorthand.cpp
//
// Created by Michal Janecek on 27.01.2024.
//

#include "Shorthand.h"

Shorthand::Shorthand(const std::string& type, const std::string& description, const std::vector<Reference>& references, const MetaBlock& metaBlock)
        : type(type), description(description), references(references), metaBlock(metaBlock) {
}


void Shorthand::deserialize(const YAML::Node& node) {
    if (!node["description"]) {
        throw std::runtime_error("Shorthand YAML node is missing a 'description'");
    }

    description = node["description"].as<std::string>();

    // Deserialize References
    if (node["references"]) {
        references.clear();  // Clear existing references before deserializing
        for (const auto& refNode : node["references"]) {
            Reference ref;
            ref.deserialize(refNode); 
            references.push_back(ref);
        }
    }

    if (node["meta_block"]) {
        metaBlock.deserialize(node["meta_block"]);  
    }
}
